collector.go

```diff
@@ -1,37 +1,35 @@
 package taskgroup
 
+import "sync"
+
 // A Collector collects values reported by task functions and delivers them to
 // an accumulator function.
 type Collector[T any] struct {
-	ch chan<- T
-	s  *Single[error]
+	μ      sync.Mutex
+	handle func(T)
+}
+
+// report delivers v to the callback under the lock.
+func (c *Collector[T]) report(v T) {
+	c.μ.Lock()
+	defer c.μ.Unlock()
+	c.handle(v)
 }
 
 // NewCollector creates a new collector that delivers task values to the
 // specified accumulator function. The collector serializes calls to value, so
-// that it is safe for the function to access shared state without a lock.  The
-// caller must call Wait when the collector is no longer needed, even if it has
-// not been used.
-func NewCollector[T any](value func(T)) *Collector[T] {
-	ch := make(chan T)
-	s := Go(NoError(func() {
-		for v := range ch {
-			value(v)
-		}
-	}))
-	return &Collector[T]{ch: ch, s: s}
-}
+// that it is safe for the function to access shared state without a lock.
+//
+// The tasks created from a collector do not return until all the values
+// reported by the underlying function have been processed by the accumulator.
+func NewCollector[T any](value func(T)) *Collector[T] { return &Collector[T]{handle: value} }
 
-// Wait stops the collector and blocks until it has finished processing.
-// It is safe to call Wait multiple times from a single goroutine.
-// Note that after Wait has been called, c is no longer valid.
-func (c *Collector[T]) Wait() {
-	if c.ch != nil {
-		close(c.ch)
-		c.ch = nil
-		c.s.Wait()
-	}
-}
+// Wait waits until the collector has finished processing.
+//
+// Deprecated: This method is now a noop; it is safe but unnecessary to call
+// it.  The state serviced by c is settled once all the goroutines writing to
```

Comment:
All the tasks, rather than goroutines, maybe? Goroutines is correct, but made me stop and go "wait, is it possible to submit stuff to a collector outside of a task?"

Hmmm, I suppose it is, since Task is just a naked function type, you could absolutely invoke one by hand as you please. Hmm. But maybe saying "once all the Tasks have returned" is still correct? Regardless of how they're invoked exactly, once all those functions return, the collector is settled.

Reply:
Yeah, that seems better, will clarify.